### PR TITLE
onboarding modal

### PR DESCRIPTION
**Pull request by tonypls**
## Issue
https://linear.app/electricitymaps/issue/ELE-1339/[ui]-onboarding-modal-component

## Description

This was a recreate task but as it's specific behaviour and quite isolated I chose to duplicate with some refactoring

### CODE REVIEW

File: web/src/components/modals/OnboardingModalInner.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
+import { useState, ReactElement } from 'react';
+
+import { useTranslation } from 'translation/translation';
+import { HiCheck, HiChevronLeft, HiChevronRight, HiXMark } from 'react-icons/hi2';
+
+interface view {
```
Comment (Kongkille) [inline web/src/components/modals/OnboardingModalInner.tsx:6]: Wondering if 'Page' is a better term?

File: web/src/utils/localStorage.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+const saveKey = (key: string, value: string) => {
+  window.localStorage.setItem(key, value);
+};
+
+const getKey = (key: string) => {
+  return window.localStorage.getItem(key);
+};
+
+export { saveKey, getKey };
```
Comment (Kongkille) [inline web/src/utils/localStorage.ts:9]: I think that we should use the atomWithStorage for consistency

https://jotai.org/docs/utils/atom-with-storage

**Review by Kongkille — COMMENTED** [review]
Looks great,

Had a few comments

File: web/src/components/modals/OnboardingModalInner.tsx (hunk 1 above)
Comment (tonypls) [reply web/src/components/modals/OnboardingModalInner.tsx]: Yeah, have updated

File: web/src/utils/localStorage.ts (hunk 2 above)
Comment (tonypls) [reply web/src/utils/localStorage.ts]: Okay. It took a lil to get it to work with booleans but should be good.

Comment (tonypls) [reply web/src/utils/localStorage.ts]: I've also renamed index.ts to atoms.ts to make it easier for search navigation

File: web/src/utils/state/atoms.ts [hunk 3]
```diff
@@ -67,3 +67,11 @@ export const isLeftPanelOpenAtom = atomWithStorage(
...
   createJSONStorage(() => sessionStorage)
 );
+
+export const hasOnboardingBeenSeenAtom = atomWithCustomStorage({
+  key: 'onboardingSeen',
+  initialValue: 'false',
+  options: {
+    syncWithLocalStorage: true,
+  },
+});
```
Comment (Kongkille) [inline web/src/utils/state/atoms.ts:77]: If we use the jotai native atomWithStorage and not my scrappy remake I think booleans work out of the box

**Review by Kongkille — APPROVED** [review]

File: web/src/utils/state/atoms.ts (hunk 3 above)
Comment (tonypls) [reply web/src/utils/state/atoms.ts:77]: Ah okay, I thought this custom one might have been preplanning for storage on the app too, not sure if the Jotai one will work for both out of the box?

File: web/src/components/modals/OnboardingModal.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    isMainTitle: true,
+    renderContent: (__: (translationKey: string) => string) => (
+      <React.Fragment>
+        <div>
+          <h1 className="text-2xl">Electricity Maps</h1>
+        </div>
+        <div className=" py-6 text-xl">
+          <h2>{__('onboarding-modal.view1.subtitle')}</h2>
+        </div>
+      </React.Fragment>
```
Comment (VIKTORVAV99) [inline web/src/components/modals/OnboardingModal.tsx:20]:
```suggestion
      <>
        <div>
          <h1 className="text-2xl">Electricity Maps</h1>
        </div>
        <div className=" py-6 text-xl">
          <h2>{__('onboarding-modal.view1.subtitle')}</h2>
        </div>
      </>
```

File: web/src/components/modals/OnboardingModal.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  },
+  {
+    headerImage: resolvePath('onboarding/mapExtract.png'),
+    renderContent: (__: (translationKey: string) => string) => (
+      <React.Fragment>
+        <div>
+          <h2 className="text-xl">{__('onboarding-modal.view2.header')}</h2>
+        </div>
+        <div>{__('onboarding-modal.view2.text')}</div>
+      </React.Fragment>
```
Comment (VIKTORVAV99) [inline web/src/components/modals/OnboardingModal.tsx:31]:
```suggestion
      <>
        <div>
          <h2 className="text-xl">{__('onboarding-modal.view2.header')}</h2>
        </div>
        <div>{__('onboarding-modal.view2.text')}</div>
      </>
```

File: web/src/components/modals/OnboardingModal.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  },
+  {
+    headerImage: resolvePath('onboarding/exchangeArrows.png'),
+    renderContent: (__: (translationKey: string) => string) => (
+      <React.Fragment>
+        <div>
+          <h2 className="text-xl">{__('onboarding-modal.view3.header')}</h2>
+        </div>
+        <div>{__('onboarding-modal.view3.text')}</div>
+      </React.Fragment>
```
Comment (VIKTORVAV99) [inline web/src/components/modals/OnboardingModal.tsx:42]:
```suggestion
      <>
        <div>
          <h2 className="text-xl">{__('onboarding-modal.view3.header')}</h2>
        </div>
        <div>{__('onboarding-modal.view3.text')}</div>
      </>
```

File: web/src/components/modals/OnboardingModal.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  },
+  {
+    headerImage: resolvePath('onboarding/splitLayers.png'),
+    renderContent: (__: (translationKey: string) => string) => (
+      <React.Fragment>
+        <div>
+          <h2 className="text-xl">{__('onboarding-modal.view4.header')}</h2>
+        </div>
+        <div>{__('onboarding-modal.view4.text')}</div>
+      </React.Fragment>
```
Comment (VIKTORVAV99) [inline web/src/components/modals/OnboardingModal.tsx:53]:
```suggestion
      <>
        <div>
          <h2 className="text-xl">{__('onboarding-modal.view4.header')}</h2>
        </div>
        <div>{__('onboarding-modal.view4.text')}</div>
      </>
```

File: web/src/utils/state/atomWithCustomStorage.ts [hunk 8]
```diff
@@ -128,6 +128,12 @@ export default function atomWithCustomStorage<Value extends string>({
       if (typeof value == typeof NO_STORAGE_VALUE) {
         return initialValue as Value;
       }
+      if (value === 'true') {
+        return true as unknown as Value;
+      }
+      if (value === 'false' || value === undefined) {
+        return false as unknown as Value;
+      }
```
Comment (VIKTORVAV99) [inline web/src/utils/state/atomWithCustomStorage.ts:136]: Is the `as unknown` needed here? Wouldn't `as Value` be enough on it's own?

File: web/src/utils/state/atoms.ts (hunk 3 above)
Comment (VIKTORVAV99) [reply web/src/utils/state/atoms.ts:77]: Don't the apps use the same interface to URLParams, localStorage and sessionStorage as the web?

**Review by VIKTORVAV99 — COMMENTED** [review]
Just a few suggestions regarding using React shorthand for fragments and some general questions.

File: web/src/utils/state/atoms.ts (hunk 3 above)
Comment (Kongkille) [reply web/src/utils/state/atoms.ts:77]: Yes it's similar, but capacitor has a few features on top which makes persisting things a bit better for native apps

https://capacitorjs.com/docs/apis/preferences

Comment (VIKTORVAV99) [reply web/src/utils/state/atoms.ts:77]: > Mobile OSs may periodically clear data set in window.localStorage, so this API should be used instead. This API will fall back to using localStorage when running as a Progressive Web App.

Makes sense why we would want to use custom atom then, but wasn't the plan to run it as a PWA? Or was that just a 4th target/enchantment for web? (Current targets: iOS, Android and Web).

File: web/src/utils/state/atomWithCustomStorage.ts (hunk 8 above)
Comment (VIKTORVAV99) [reply web/src/utils/state/atomWithCustomStorage.ts:136]: Ah okay, that makes sense. I think we should try and create a custom interface for this later.

File: web/src/components/modals/OnboardingModalInner.tsx [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  headerImage: { pathname: string };
+  isMainTitle?: boolean;
+  renderContent: (argument0: any) => ReactElement;
+  title?: (argument0: any) => string;
```
[the remark this answers is not in the document]
Comment (tonypls) [reply web/src/components/modals/OnboardingModalInner.tsx]: I've made these more descriptive :D

**Pull request merged**